core/Application/Environment.php

```diff
@@ -0,0 +1,176 @@
+use Piwik\Application\Kernel\GlobalSettingsProvider;
+use Piwik\Application\Kernel\PluginList;
+use Piwik\Application\Kernel\PluginList\IniPluginList;
+use Piwik\Application\Kernel\GlobalSettingsProvider\IniSettingsProvider;
+use Piwik\Container\ContainerFactory;
+use Piwik\Container\StaticContainer;
+use Piwik\Piwik;
+
+/**
+ * Encapsulates Piwik environment setup and access.
+ *
+ * The Piwik environment consists of two main parts: the kernel and the DI container.
+ *
+ * The 'kernel' is the core part of Piwik that cannot be modified / extended through the DI container.
+ * It includes components that are required to create the DI container.
+ *
+ * Currently the only objects in the 'kernel' are a GlobalSettingsProvider object and a
+ * PluginList object. The GlobalSettingsProvider object is required for the current PluginList
+ * implementation and for checking whether Development mode is enabled. The PluginList is
+ * needed in order to determine what plugins are activated, since plugins can provide their
+ * own DI configuration.
+ *
+ * The DI container contains every other Piwik object, including the Plugin\Manager,
+ * plugin API instances, dependent services, etc. Plugins and users can override/extend
+ * the objects in this container.
+ *
+ * NOTE: DI support in Piwik is currently a work in process; not everything is currently
+ * stored in the DI container, but we are working towards this.
+ */
+class Environment
+{
+    /**
+     * @var string
+     */
+    private $environment;
+
+    /**
+     * @var array
+     */
+    private $definitions;
+
+    /**
+     * @var Container
+     */
+    private $container;
+
+    /**
+     * @var GlobalSettingsProvider
+     */
+    private $globalSettingsProvider;
+
+    /**
+     * @var PluginList
+     */
+    private $pluginList;
+
+    /**
+     * @param string $environment
+     * @param array $definitions
+     */
+    public function __construct($environment, array $definitions = array())
+    {
+        $this->environment = $environment;
+        $this->definitions = $definitions;
+    }
+
+    /**
+     * Initializes the kernel globals and DI container.
+     */
+    public function init()
+    {
+        $this->container = $this->createContainer();
+
+        StaticContainer::set($this->container);
+
+        $this->validateEnvironment();
+
+        Piwik::postEvent('Environment.bootstrapped'); // this event should be removed eventually
+    }
+
+    /**
+     * Returns the DI container. All Piwik objects for a specific Piwik instance should be stored
+     * in this container.
+     *
+     * @return Container
+     */
+    public function getContainer()
+    {
+        return $this->container;
+    }
+
+    /**
+     * Returns the Piwik object for this Environment's Piwik instance by ID.
+     *
+     * @param string $id
+     * @return mixed
+     * @throws \DI\NotFoundException
+     */
+    public function get($id)
+    {
+        return $this->container->get($id);
```

Comment:
Sorry I just realized you pinged me twice in the issue but somehow I never saw that.

Regarding what you said above I would rather separate the concerns clearly and have the container be used through `ContainerInterface` rather than through the Environment class. I.e. I don't see a gain to add another way to use the container?

```php
$instanceManager = ...
$instanceManager->forEach(function (ContainerInterface $c) {
    $api = $c->get('...\API');
    $api->doSomething();
});
```

Or:

```php
$container = ...
$instanceManager = ...
$instanceManager->forEach(function () use ($container) {
    $api = $container->get('...\API');
    $api->doSomething();
});
```